diff --git a/packages/opencode/src/plugin/github-copilot/copilot.ts b/packages/opencode/src/plugin/github-copilot/copilot.ts
index 6f0e46402..de3b78db8 100644
--- a/packages/opencode/src/plugin/github-copilot/copilot.ts
+++ b/packages/opencode/src/plugin/github-copilot/copilot.ts
@@ -97,6 +97,12 @@ export async function CopilotAuthPlugin(input: PluginInput): Promise<Hooks> {
             const { isVision, isAgent } = iife(() => {
               try {
                 const body = typeof init?.body === "string" ? JSON.parse(init.body) : init?.body
+                const userMsgsCount = Math.max(1, (body?.messages ?? body?.input ?? [])
+                  .map((x: { role?: string, content?: string | any[] }) => ((x?.role === "user" &&
+                    (Array.isArray(x?.content) ? x.content.some((part: any) => part?.type !== "tool_result") : !/^[/@]/.test(x?.content as string))) ? 1 : 0))
+                  .reduce((acc: number, x: number) => acc + x, 0))
+                const notUser = userMsgsCount > 1
+                // const notUser = userMsgsCount % 10 !== 1
 
                 // Completions API
                 if (body?.messages && url.includes("completions")) {
@@ -106,7 +112,7 @@ export async function CopilotAuthPlugin(input: PluginInput): Promise<Hooks> {
                       (msg: any) =>
                         Array.isArray(msg.content) && msg.content.some((part: any) => part.type === "image_url"),
                     ),
-                    isAgent: last?.role !== "user" || imgMsg(last),
+                    isAgent: last?.role !== "user" || imgMsg(last) || notUser,
                   }
                 }
 
@@ -118,7 +124,7 @@ export async function CopilotAuthPlugin(input: PluginInput): Promise<Hooks> {
                       (item: any) =>
                         Array.isArray(item?.content) && item.content.some((part: any) => part.type === "input_image"),
                     ),
-                    isAgent: last?.role !== "user" || imgMsg(last),
+                    isAgent: last?.role !== "user" || imgMsg(last) || notUser,
                   }
                 }
 
@@ -140,10 +146,10 @@ export async function CopilotAuthPlugin(input: PluginInput): Promise<Hooks> {
                               part.content.some((nested: any) => nested?.type === "image")),
                         ),
                     ),
-                    isAgent: !(last?.role === "user" && hasNonToolCalls) || imgMsg(last),
+                    isAgent: !(last?.role === "user" && hasNonToolCalls) || imgMsg(last) || notUser,
                   }
                 }
-              } catch {}
+              } catch { }
               return { isVision: false, isAgent: false }
             })
 
diff --git a/packages/opencode/src/provider/provider.ts b/packages/opencode/src/provider/provider.ts
index 9aa1b6304..85cae8c9d 100644
--- a/packages/opencode/src/provider/provider.ts
+++ b/packages/opencode/src/provider/provider.ts
@@ -188,6 +188,15 @@ function custom(dep: CustomDep): Record<string, CustomLoader> {
         },
         options: {},
       }),
+    occo: () =>
+      Effect.succeed({
+        autoload: true,
+        async getModel(sdk: any, modelID: string, _options?: Record<string, any>) {
+          if (useLanguageModel(sdk)) return sdk.languageModel(modelID)
+          return shouldUseCopilotResponsesApi(modelID) ? sdk.responses(modelID) : sdk.chat(modelID)
+        },
+        options: {},
+      }),
     "github-copilot": () =>
       Effect.succeed({
         autoload: false,
@@ -1614,7 +1623,7 @@ const layer: Layer.Layer<
       if (providerID.startsWith("opencode")) {
         priority = ["gpt-5-nano"]
       }
-      if (providerID.startsWith("github-copilot")) {
+      if (providerID.startsWith("github-copilot") || providerID.startsWith("occo")) {
         priority = ["gpt-5-mini", "claude-haiku-4.5", ...priority]
       }
       for (const item of priority) {
